query-engine/connectors/query-connector/src/filter/scalar.rs

```diff
@@ -116,58 +123,191 @@ impl ScalarFilter {
 
                 inner(list, chunk_size)
                     .into_iter()
                     .map(|batch| ScalarFilter {
                         projection: projection.clone(),
-                        condition: ScalarCondition::In(batch),
+                        condition: ScalarCondition::In(ConditionListValue::list(batch)),
                         mode: mode.clone(),
                     })
                     .collect()
             }
 
-            ScalarCondition::NotIn(list) => {
+            ScalarCondition::NotIn(ConditionListValue::List(list)) => {
                 let projection = self.projection;
 
                 inner(list, chunk_size)
                     .into_iter()
                     .map(|batch| ScalarFilter {
                         projection: projection.clone(),
-                        condition: ScalarCondition::NotIn(batch),
+                        condition: ScalarCondition::NotIn(ConditionListValue::list(batch)),
                         mode: mode.clone(),
                     })
                     .collect()
             }
             _ => vec![self],
         }
     }
+
+    /// Returns the referenced scalar field if there is one.
+    pub fn as_field_ref(&self) -> Option<&ScalarFieldRef> {
+        self.condition.as_field_ref()
+    }
+
+    /// Returns all the scalar fields related to a scalar filter.
+    /// It also includes the referenced field if there is one.
+    pub fn scalar_fields(&self) -> Vec<&ScalarFieldRef> {
+        let mut fields = self.projection.scalar_fields();
+
+        if let Some(field_ref) = self.as_field_ref() {
+            fields.push(field_ref);
+        }
+
+        fields
+    }
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum ScalarCondition {
-    Equals(PrismaValue),
-    NotEquals(PrismaValue),
-    Contains(PrismaValue),
-    NotContains(PrismaValue),
-    StartsWith(PrismaValue),
-    NotStartsWith(PrismaValue),
-    EndsWith(PrismaValue),
-    NotEndsWith(PrismaValue),
-    LessThan(PrismaValue),
-    LessThanOrEquals(PrismaValue),
-    GreaterThan(PrismaValue),
-    GreaterThanOrEquals(PrismaValue),
-    In(PrismaListValue),
-    NotIn(PrismaListValue),
+    Equals(ConditionValue),
+    NotEquals(ConditionValue),
+    Contains(ConditionValue),
+    NotContains(ConditionValue),
+    StartsWith(ConditionValue),
+    NotStartsWith(ConditionValue),
+    EndsWith(ConditionValue),
+    NotEndsWith(ConditionValue),
+    LessThan(ConditionValue),
+    LessThanOrEquals(ConditionValue),
+    GreaterThan(ConditionValue),
+    GreaterThanOrEquals(ConditionValue),
+    In(ConditionListValue),
+    NotIn(ConditionListValue),
     JsonCompare(JsonCondition),
-    Search(PrismaValue, Vec<ScalarProjection>),
-    NotSearch(PrismaValue, Vec<ScalarProjection>),
+    Search(ConditionValue, Vec<ScalarProjection>),
+    NotSearch(ConditionValue, Vec<ScalarProjection>),
     IsSet(bool),
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ConditionValue {
```

Comment:
Yep, we can do that. Here's my question though: to what extent would you split? Cause there are a lot of things in this file. If I look at all symbols we have
- `ScalarFilter`
- `QueryMode`
- `ScalarCondition`
- `ScalarProjection`
- `ConditionValue`
- `ConditionListValue`

What would be your ideal file tree for these structs? Here's one I could imagine:
- `scalars/mod.rs` -> `ScalarFilter` + `QueryMode` (since it's a very simple enum)
- `scalars/projection.rs` -> `ScalarProjection`
- `scalars/condition/mod.rs` -> `ScalarCondition`
- `scalars/condition/value.rs` -> `ConditionValue` + `ConditionListValue`

Feel free to suggest anything you prefer 👍

Reply:
done here [ba9f88c](https://github.com/prisma/prisma-engines/pull/3088/commits/ba9f88c012a87a9ab6de3b5eb248ba20b4cdfd15)